File: 算法竞赛进阶指南/0x00基本算法/AcWing94.cpp
#include <iostream>
#include <vector>
using namespace std;
int n;
int order[20];
bool chosen[20];

void calc(int k) {
  if (k == n+1) {
    for (int i=1; i<=n; i++)
      printf("%d ", order[i]);
    puts("");
    return;
  }
  for (int i=1; i<=n; i++) {
    if (chosen[i]) continue;
    order[k] = i;
    chosen[i] = 1;
    calc(k+1);
    chosen[i] = 0;
    order[k] = 0;
  }
}

int main() {
  cin >> n;
  calc(1);
  return 0;
}

File: 0x00基本算法/0x02递推与递归/AcWing94.py
n, order, chosen = 0, [0]*20, [0]*20
def calc(k):
    if k == n+1:
        for i in range(1, n+1):
            print('%d ' % order[i], end='')
        print()
        return
    for i in range(1, n+1):
        if chosen[i]: continue
        order[k] = i
        chosen[i] = 1
        calc(k+1)
        chosen[i] = 0
        order[k] = 0

def main():
    global n
    n = int(input())
    calc(1)
    
main()